### PR TITLE
[SYSTEMDS-200] Gaussian Mixture Model builtin

### PR DESCRIPTION
**Pull request by Shafaq-Siddiqi**
New Builtin for Gaussian Mixture Model with four covariance types namely VVV, EEE, diag and spherical (please see the script header for more details).

Spark tests for using kmeans as initialization function are commented out due to a runtime exception in kmeans  execution on the following instruction. 

SPARK°rexpand°cast=true°max=_Var872°ignore=false°dir=cols°target=_mVar884°_mVar885·MATRIX·FP64

### CODE REVIEW

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinGMMTest.java [hunk 1]
```diff
@@ -44,69 +45,66 @@
...
+	}
+
+	@Test public void testGMMM4KmeanSpark() {
+		runGMMTest(3, "VII", "kmeans", 100, 0.000001, false, LopProperties.ExecType.SPARK);
+	}
+
+	private void runGMMTest(int G_mixtures, String model, String init_param, int iter, double reg, boolean rewrite,
 			LopProperties.ExecType instType) {
 		Types.ExecMode platformOld = setExecMode(instType);
+		OptimizerUtils.ALLOW_ALGEBRAIC_SIMPLIFICATION = rewrite;
```
Comment (Shafaq-Siddiqi) [inline src/test/java/org/apache/sysds/test/functions/builtin/BuiltinGMMTest.java:107]: Spark test passed after setting OptimizerUtils.ALLOW_ALGEBRAIC_SIMPLIFICATION to false.

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinGMMTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	@Test public void testGMMM2KmeanSpark() {
+		runGMMTest(3, "EEE", "kmeans", 100, 0.000001, false, LopProperties.ExecType.SPARK);
+	}
+
+	@Test public void testGMMM3KmeanSpark() {
+		runGMMTest(3, "VVI", "kmeans", 100, 0.000001, false, LopProperties.ExecType.SPARK);
+	}
+
+	@Test public void testGMMM4KmeanSpark() {
+		runGMMTest(3, "VII", "kmeans", 100, 0.000001, false, LopProperties.ExecType.SPARK);
```
[the remark this answers is not in the document]
Comment (Shafaq-Siddiqi) [reply src/test/java/org/apache/sysds/test/functions/builtin/BuiltinGMMTest.java:105]: Yes, We can switch them. I will push a commit.

File: src/test/scripts/functions/transform/input/iris/iris.csv [hunk 3]
```diff
@@ -148,4 +148,4 @@ ID,Sepal.Length,Sepal.Width,Petal.Length,Petal.Width,Species
 147,6.3,2.5,5,1.9,virginica
 148,6.5,3,5.2,2,virginica
 149,6.2,3.4,5.4,2.3,virginica
-150,5.9,3,5.1,1.8,virginica
+150,5.9,3,5.1,1.8,virginica
```
[the remark this answers is not in the document]
Comment (Shafaq-Siddiqi) [reply src/test/scripts/functions/transform/input/iris/iris.csv:151]: I just find that unnecessary and removed it.

File: scripts/builtin/gmm.dml [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+# INPUT PARAMETERS:
+# ---------------------------------------------------------------------------------------------
+# NAME            TYPE    DEFAULT     MEANING
+# ---------------------------------------------------------------------------------------------
+# X               Double   ---       Matrix X  
+# n_components    Integer  3         Number of n_components in the Gaussian mixture model
+# model           String   "VVV"     "VVV": unequal variance (full),each component has its own general covariance matrix
+#                                    "EEE": equal variance (tied), all components share the same general covariance matrix
+#                                    "VVI": spherical, unequal volume (diag), each component has its own diagonal covariance matrix 
+#                                    "VII": spherical, equal volume (spherical), each component has its own single variance
```
[the remark this answers is not in the document]
Comment (Shafaq-Siddiqi) [reply scripts/builtin/gmm.dml:35]: Yes, these are model types and saying GMM-VVV (R format) or GMM with "full" covariance (Python format) is fine.

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinGMMTest.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import org.junit.Test;
+
+import java.util.HashMap;
+
+public class BuiltinGMMTest extends AutomatedTestBase {
+	private final static String TEST_NAME = "GMM";
+	private final static String TEST_DIR = "functions/builtin/";
+	private static final String TEST_CLASS_DIR = TEST_DIR + BuiltinGMMTest.class.getSimpleName() + "/";
+
+	private final static double eps = 1;
```
[the remark this answers is not in the document]
Comment (Shafaq-Siddiqi) [reply src/test/java/org/apache/sysds/test/functions/builtin/BuiltinGMMTest.java:38]: The test is comparing the number of outliers which could be any number greater than equal to zero.

File: scripts/builtin/gmm.dml [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+    log_prob = (bc_matrix * t(rowSums(mu^2) * precisions) -
+                    2 * X %*% t(mu * precisions) +
+                    rowSums(X*X) %*% t(precisions) ) # TODO replace rowSums(X*X) with squared rowNorm() built-in
+  }
+  if(ncol(log_det) == 1)
+    log_det = matrix(1, 1, ncol(log_prob)) * log_det 
+  es_log_prob = -.5 * (d * log(2 * pi) + log_prob) + log_det
+}
+
+logsumexp = function(Matrix[Double] M) # TODO replace with a built-in function logsumexp
```
[the remark this answers is not in the document]
Comment (Shafaq-Siddiqi) [reply scripts/builtin/gmm.dml:344]: I am planning to do it in few days.

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinGMMTest.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	private final static double tol = 1e-3;
+	private final static int rows = 100;
+	private final static double spDense = 0.99;
+	private final static String DATASET = SCRIPT_DIR + "functions/transform/input/iris/iris.csv";
+
+	@Override public void setUp() {
+		addTestConfiguration(TEST_NAME, new TestConfiguration(TEST_CLASS_DIR, TEST_NAME, new String[] {"B"}));
+	}
+
+	@Test public void testGMMM1() { runGMMTest(3, "VVV", "random", 100, 0.000001, true, LopProperties.ExecType.CP); }
```
[the remark this answers is not in the document]
Comment (Shafaq-Siddiqi) [reply src/test/java/org/apache/sysds/test/functions/builtin/BuiltinGMMTest.java:48]: > Hi @Shafaq-Siddiqi,
> 
> I think this looks awesome, 🥇
> Unfortunately I'm not that well versed in the internals of GMM so i hope the math is in order, but looking at the internal operations i really look forward to try it out on the compressed representation!
> 
> I have some minor comments and questions you that would be great if you could answer!
> 
> On the less important side, the script `gmm.dml` as it is now does not follow what i thought to be the coding guidelines.
> 
> 1. Variable inputs names should be CamelCase.
> 2. Indentation should be tabs not double space.
> 
> I'm fine if you don't want to change it, but i really value if we could aim for some consistency between our scripts. I can also change my formatting to the settings you use.
> 
> I also can't help but notice that there are no doc entry added for this new operation, that would also be great!

If I remember correctly, we agreed on two spaces in dml. I will confirm and then will change accordingly.

File: scripts/builtin/gmm.dml [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+# ------------------------------------------
+# Gaussian Mixture Model
+# ------------------------------------------
+
+# INPUT PARAMETERS:
+# ---------------------------------------------------------------------------------------------
+# NAME            TYPE    DEFAULT     MEANING
+# ---------------------------------------------------------------------------------------------
+# X               Double   ---       Matrix X  
+# n_components    Integer  3         Number of n_components in the Gaussian mixture model
```
Comment (OlgaOvcharenko) [inline scripts/builtin/gmm.dml:31]: Default n_components value is 1 in line 55.

**Comment by Shafaq-Siddiqi** [comment]
Merged into Master.

**Pull request closed without merging**